Add documentation for string formats; fix `at-identifier` format

zzstoatzz opened the pull request:
closes https://github.com/MarshalX/atproto/issues/478

this PR adds:
- [docstrings](https://atproto--491.org.readthedocs.build/en/491/atproto/atproto_client.models.string_formats.html) to all validator functions
- a new docs [page, `string_formats.md`](https://atproto--491.org.readthedocs.build/en/491/atproto_client/string_formats.html)

- improved rendering of string format contstraints, as [discussed below](https://github.com/MarshalX/atproto/pull/491#issuecomment-2544734404)

MarshalX commented on packages/atproto_client/models/string_formats.py:
```diff
@@ -1,3 +1,14 @@
+
+- https://github.com/bluesky-social/atproto/tree/main/interop-test-files/syntax
+
+see https://github.com/MarshalX/atproto/pull/406 for more details.
```
could we create string_formats.md (same as we have, for examle, for `auth.md` and write all the details? instead of mentioning Pull Request. 

also it would be great to provide here a little code snippet how to use this optonal validation; as we have in the main readme, for example

MarshalX commented on packages/atproto_client/models/string_formats.py:
```diff
@@ -52,19 +63,49 @@
+    :param validate_fn: The validation function to conditionally apply
+    :type validate_fn: Callable[..., str]
+    :return: A wrapped validation function that only validates in strict mode
+    :rtype: Callable[..., str]
```
the entire project is documented in Google Docstring Style; could you pls align it? here is the guide: https://google.github.io/styleguide/pyguide.html#38-comments-and-docstrings (or just see how other paths of the SDK is documented)

MarshalX reviewed (commented):
thank you! could you pls investigate how to fix this rendering (screnshot)? happened after merging PR with string formats 

<img width="1213" alt="image" src="https://github.com/user-attachments/assets/cb727fb4-ab2e-4ad1-97c6-0a602415e4d5" />

link: https://atproto--491.org.readthedocs.build/en/491/atproto/atproto_client.models.app.bsky.graph.get_follows.html#atproto_client.models.app.bsky.graph.get_follows.Params.actor

zzstoatzz commented:
@MarshalX alright I have

- updated the docstrings to follow google convention
- added a `string_formats.md` (let me know what you think!)
- explored the weird `<locals>.wrapper` rendering a bit more

on the last point, the best way I could immediately find to deal with this was to switch to using `@wraps(validate_fn)` in `only_validate_if_strict`, since that properly handles carrying metadata without manually patching `__doc__` or `__name__`. However, as a consequence this forces me to update each validator signature to at least `(v: str, _: ValidationInfo)`. With this in place, we see a more reasonable rendering

<img width="581" alt="image" src="https://github.com/user-attachments/assets/b9f6759a-6ed3-422f-9aa4-306d585ec285" />


while the address present is midly annoying to me, showing the function name makes some sense to me as the constraint really is arbitrary logic that may exist in that function. let me know if you see a more reasonable path forward on this!

zzstoatzz commented on packages/atproto_client/models/string_formats.py:
```diff
@@ -54,6 +54,19 @@
 )
 
 
+class _NamedValidator:
+    """Decorator to add a __str__ attribute to a validation function."""
+
+    def __init__(self, validate_fn: Callable[..., str]) -> None:
+        self.validate_fn = validate_fn
+
+    def __call__(self, v: str, info: ValidationInfo) -> str:
+        return self.validate_fn(v, info)
+
+    def __str__(self) -> str:
+        return f'Validated by: {self.validate_fn.__name__} (only when `strict_string_format=True`)'
```
doh! i tried almost exactly this but __repr__ instead and it broke the docstrings in this file

zzstoatzz commented:
great @MarshalX ! works for me

MarshalX commented:
@zzstoatzz i fixed some stuff; pls check and confirm is all good

zzstoatzz reviewed (commented):
nice! lgtm

MarshalX commented:
thank you!
